### PR TITLE
BUG: Fix HDFStore empty keys on native HDF5 file by adding keyword include

### LINKED ISSUE #29916: HDF5: empty groups and keys
Hi,

With some of the hdf5 files I have, `pandas.HDFStore.groups()` returns an empty list. (as does `.keys()` which iterates over the groups). However, the data are accessible via `.get()` or `.get_node()`.

This is related to #21543 and #21372 where the `.groups()` logic was changed, in particular using `self._handle.walk_groups()` instead of `self._handle.walk_nodes()`, now to be found here:
https://github.com/pandas-dev/pandas/blob/ea2e26ae7d700d7fd363ea5bfc05d2fe3fb8a5ee/pandas/io/pytables.py#L1212

#### Current Output

```python
>>> hdf.groups()
[]
```
```python
>>> hdf.keys()
[]
```

#### Expected Ouptut

List of groups and keys as visible with e.g. `h5dump`.
**Note:** Changing the aforementioned line back to use `.walk_nodes()` fixes the issue and lists the groups and keys properly:

```python
>>> hdf.groups()
[/Data/Table Layout (Table(69462,), zlib(4)) ''
   description := {
...
/Data/Array Layout/2D Parameters/Data Parameters (Table(15,)) ''
   description := {
   "mnemonic": StringCol(itemsize=8, shape=(), dflt=b'', pos=0),
   "description": StringCol(itemsize=48, shape=(), dflt=b'', pos=1),
   "isError": Int64Col(shape=(), dflt=0, pos=2),
   "units": StringCol(itemsize=7, shape=(), dflt=b'', pos=3),
   "category": StringCol(itemsize=31, shape=(), dflt=b'', pos=4)}
   byteorder := 'little'
   chunkshape := (642,)]]
```
```python
>>> hdf.keys()
['/Data/Table Layout',
 '/Metadata/Data Parameters',
 '/Metadata/Experiment Notes',
 '/Metadata/Experiment Parameters',
 '/Metadata/Independent Spatial Parameters',
 '/Metadata/_record_layout',
 '/Data/Array Layout/Layout Description',
 '/Data/Array Layout/1D Parameters/Data Parameters',
 '/Data/Array Layout/2D Parameters/Data Parameters']
```

#### Fix

One solution would be (I guess) to revert #21543, another to fix at least `.keys()` to use `._handle.walk_nodes()` instead of `.groups()` in
https://github.com/pandas-dev/pandas/blob/ea2e26ae7d700d7fd363ea5bfc05d2fe3fb8a5ee/pandas/io/pytables.py#L562

Could also be that it is a bug in `pytables`.

#### Problem background

I was trying to figure out why some hdf5 files open fine with `pandas` but fail with `dask`.
The reason is that `dask` allows wildcards and iterates over the keys to find valid ones. If `.keys()` is empty, reading the files with `dask` fails.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 3.10.0-957.27.2.el7.x86_64
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : C
LOCALE           : en_US.UTF-8

pandas           : 0.25.3
numpy            : 1.17.3
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 42.0.1.post20191125
Cython           : None
pytest           : 5.0.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.2
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.10.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.2
matplotlib       : 3.1.2
numexpr          : 2.7.0
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.2
sqlalchemy       : None
tables           : 3.6.1
xarray           : 0.14.1
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by roberthdevries**
- [x] closes #29916
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/pytables.py [hunk 1]
```diff
@@ -580,16 +580,37 @@ def __enter__(self):
...
+
+        Raises
+        ------
+        raises ValueError if kind has an illegal value
         """
-        return [n._v_pathname for n in self.groups()]
+        if include == "pandas":
+            return [n._v_pathname for n in self.groups()]
+
+        if include == "native":
```
Comment (jreback) [inline pandas/io/pytables.py]: use elif

File: pandas/tests/io/pytables/test_store.py [hunk 2]
```diff
@@ -296,6 +296,37 @@ def test_keys(self, setup_path):
...
+            with tables.open_file(path, mode="w") as h5file:
+                group = h5file.create_group("/", "group")
+                h5file.create_table(group, "table1", Table1, "Table 1")
+                h5file.create_table(group, "table2", Table2, "Table 2")
+                h5file.create_table(group, "table3", Table3, "Table 3")
+            with HDFStore(path) as store:
+                assert len(store.keys(include="native")) == 3
+                expected = {"/group/table1", "/group/table2", "/group/table3"}
+                assert set(store.keys(include="native")) == expected
+                assert set(store.keys(include="pandas")) == set()
```
Comment (jreback) [inline pandas/tests/io/pytables/test_store.py]: can you also assert that ``.get()`` works here

File: pandas/io/pytables.py [hunk 3]
```diff
@@ -580,16 +580,37 @@ def __enter__(self):
...
         self.close()
 
-    def keys(self) -> List[str]:
+    def keys(self, include: Optional[str] = "pandas") -> List[str]:
         """
         Return a list of keys corresponding to objects stored in HDFStore.
 
+        Parameters
+        ----------
+        include : str, default 'pandas'
```
Comment (jreback) [inline pandas/io/pytables.py]: can you add a versionadded 1.1 tag

File: pandas/io/pytables.py [hunk 4]
```diff
@@ -580,16 +580,37 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_value, traceback):
         self.close()
 
-    def keys(self) -> List[str]:
+    def keys(self, include: Optional[str] = "pandas") -> List[str]:
```
Comment (jreback) [inline pandas/io/pytables.py]: this is not Optional, just str (alternatively, we could actually do ```Optiona[str] = None``` and then allow passing include='native')

but find either way

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine. does this solve the OP? meaning folks simply want to use .keys() to get *all* the tables, then use .get() to read them?

File: pandas/io/pytables.py (hunk 1 above)
Comment (roberthdevries) [reply pandas/io/pytables.py]: Not really needed in this case as the previous if returns in all cases. Why would that help us?

File: pandas/io/pytables.py (hunk 3 above)
Comment (roberthdevries) [reply pandas/io/pytables.py]: added

File: pandas/io/pytables.py (hunk 4 above)
Comment (roberthdevries) [reply pandas/io/pytables.py]: Removed the `Optional` annotation

File: pandas/tests/io/pytables/test_store.py (hunk 2 above)
Comment (roberthdevries) [reply pandas/tests/io/pytables/test_store.py]: added a test

File: pandas/io/pytables.py [hunk 5]
```diff
@@ -580,16 +580,40 @@ def __enter__(self):
...
 
-    def keys(self) -> List[str]:
+    def keys(self, include: str = "pandas") -> List[str]:
         """
         Return a list of keys corresponding to objects stored in HDFStore.
 
+        Parameters
+        ----------
+
+        .. versionadded:: 1.1.0
```
Comment (WillAyd) [inline pandas/io/pytables.py]: Same comment on other PR - put below include and indent one level (I think will fix CI)

Comment (roberthdevries) [reply pandas/io/pytables.py]: fixed

**Comment by roberthdevries** [comment]
> looks fine. does this solve the OP? meaning folks simply want to use .keys() to get _all_ the tables, then use .get() to read them?

We could also merge the two sets of tables. I think pandas does not use native HDF5 tables to store DataFrames. In that case the two sets are disjoint and we should be fine returning the joined set.

**Comment by roberthdevries** [comment]
I tried the variant with returning the combination of keys of pandas tables and HDF5 native tables, and the unit test `pandas/tests/io/pytables/test_store.py::TestHDFStore::test_copy` fails. I think the solution we have now helps the `dask` developers at least to have a work-around solution, that can work on both older and newer pandas versions by checking for the pandas version.

**Comment by st-bender** [comment]
> I tried the variant with returning the combination of keys of pandas tables and HDF5 native tables, and the unit test `pandas/tests/io/pytables/test_store.py::TestHDFStore::test_copy` fails.

Hi there, this is interesting, how did this test pass before the `keys()` method changed, i.e. it used to return all the keys in pandas < 0.24?

> I think the solution we have now helps the `dask` developers at least to have a work-around solution, that can work on both older and newer pandas versions by checking for the pandas version.

Hopefully, thanks for working on it.

File: pandas/io/pytables.py (hunk 1 above)
Comment (jreback) [reply pandas/io/pytables.py:609]: can you change this (its more idiomatic)

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine, pls merge master and ping on green

File: pandas/io/pytables.py (hunk 1 above)
Comment (roberthdevries) [reply pandas/io/pytables.py]: fixed

File: pandas/io/pytables.py [hunk 6]
```diff
@@ -580,16 +580,40 @@ def __enter__(self):
...
         """
         Return a list of keys corresponding to objects stored in HDFStore.
 
+        Parameters
+        ----------
+
+        include : str, default 'pandas'
+                When kind equals 'pandas' return pandas objects
+                When kind equals 'native' return native HDF5 Table objects
+                Otherwise fail with a ValueError
```
Comment (jreback) [inline pandas/io/pytables.py:593]: this line is not needed (the otherwise fail)

**Review by jreback — CHANGES_REQUESTED** [review]
don't you need this *also* in `.groups()`?

File: pandas/io/pytables.py (hunk 6 above)
Comment (roberthdevries) [reply pandas/io/pytables.py]: Removed

**Comment by roberthdevries** [comment]
> don't you need this _also_ in `.groups()`?

Could be added there as well, although that use case is less typical. Most users are interested in the _tables_ in the file.
However, if this question pops up, I'll be happy to create a PR for that.

**Comment by roberthdevries** [comment]
@jreback tests should have run fine, except that one of the tests took too long I think. So this is my ping on sort of green

**Comment by jreback** [comment]
thanks @roberthdevries

**Comment by jamesbooker** [comment]
thanks @roberthdevries and @jreback for resolving this issue for keys, it will help my project a lot.

@roberthdevries the source files coming into my system use groups (not created by pandas) so I would need the same fix applying to .groups() as well as .keys() if at all possible.

**Pull request merged**